docs/vm-configuration.md

```diff
@@ -266,3 +266,227 @@ the generation of multiple libvirt XML documents, the domain XML and associated
 documents for secrets, network filters and node devices. So there is not a strict
 1-1 mapping between the k8s API resource representation and the libvirt low
 level representation.
+
+Mapping rules for mapping libvirt XML to the KubeVirt API
+---------------------------------------------------------
+
+In order to support libvirt features, we tend to expose the libvirt API as much
+as possible in a 1:1 fashion. However, there are some differences which need to
+be taken into account:
+
+ 1. Our API is json/yaml and not xml.
+ 2. Host specifics are not exposed on the API.
+ 3. We want to make use of the golang xml/yaml/json mappers as much as possible.
+ 4. Ordering of devices can be important (e.g. for pci address assignment).
+ 5. KubeVirt also delivers specialized device types (mostly for network and
+    storage), which need to be fit into the existing structure.
+
+**Note**: Point **4** is still under debate.
+
+## General mapping rules
+
+### All devices need to live directly under *devices*
+
+In order to allow arbitrary ordering of devices, they all need directly live
+under `devices`.
+
+This is OK:
+
+```yaml
+devices:
+  lun:
+    my: data
+  video:
+    qxl:
+      my: data
+  lun:
+    my: data1
+  video:
+    qxl:
+      my: data1
+```
+
+This is **not** OK:
+
+```yaml
+devices:
+  disks:
+    lun:
+      my: data
+    lun:
+      my: data1
+  videos:
+    qxl:
+      my: data
+    qxl:
+      my: data1
+```
+
+### Add clear extension points where KubeVirt provides specializations
+
+Disks and network interfaces are good examples where KubeVirt provides
+specializations. See the mapping decisions for them below to get an impression
+on how clear extension points can be added. In general allowing to choose
+between different optional source or target structs is the preferred mechanism,
+to cleanly separate between KubeVirt extensions and libvirt source or target
+types.
+
+For device types where no KubeVirt specific extensions are expected, creating
+sub-structs is not required. Even differentiating based on `type` fields is OK
+there. Examples are `smartcard`, `input` and `controller`.
+
+### Use optional structs to represent different sources or targets
+
+By having an inlinded wrapper-struct, containing optional structs corresponding
+to different types, we get a reasonable usable API and can maintain explicit
+mapping withing the code.
+
+Given a `Disk` which can have different `DiskSource`s
+
+```golang
```

Comment:
Do I understand correctly, that we encourage users to define their own structs?
This can lead to a disastrous mess. The basis has to be defined by our system and not by users hacking source codes of kubevirt.

Reply:
We create all the structs. This is a guide, which should help engineers modelling the structs, by providing some rules and explanations.

Reply:
Ah, thank you for clarification. I was not sure.